File: src/main/java/cubesrv/CServer.java
package cubesrv;

import static cubesrv.CubeDefs.*;
import static cubesrv.CubeRead.*;
import static cubesrv.CubeSearch.*;
import static cubesrv.ThreadPoolHelper.*;
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpHandler;
import com.sun.net.httpserver.HttpServer;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.util.concurrent.locks.ReentrantLock;

public class CServer {

    private static class SocketChunkedResponder extends Responder {
        private final OutputStream m_fdReq;
        private final ReentrantLock respLock = new ReentrantLock();

        SocketChunkedResponder(OutputStream fdReq) { m_fdReq = fdReq; }

        @Override
        public void handleMessage(MessageType mt, String msg) {
            String prefix = switch (mt) {
                case MT_UNQUALIFIED -> "";
                case MT_PROGRESS    -> "progress: ";
                case MT_MOVECOUNT   -> "moves: ";
                case MT_SOLUTION    -> "solution:";
            };
            String respBuf = prefix + msg + "\n";
            respLock.lock();
            try {
                m_fdReq.write(respBuf.getBytes());
                m_fdReq.flush();
            } catch (IOException e) {
                System.out.println("error: socket write in responder");
            } finally {
                respLock.unlock();
            }
        }
    }

    private static final ReentrantLock solverLock = new ReentrantLock();

    private static void processCubeReq(HttpExchange exchange, CubeSearcher cubeSearcher) throws IOException {
        String query = exchange.getRequestURI().getQuery();
        if (query.startsWith("stop")) {
            ProgressBase.requestStop();
            exchange.sendResponseHeaders(200, -1);
            return;
        }
        if (solverLock.tryLock()) {
            try {
                ProgressBase.requestRestart();
                exchange.getResponseHeaders().add("Content-type", "text/plain");
                exchange.sendResponseHeaders(200, 0);
                try (OutputStream fdReq = exchange.getResponseBody()) {
                    SocketChunkedResponder responder = new SocketChunkedResponder(fdReq);
                    String mode = query.substring(0, 1);
                    if (query.length() > 1 && query.charAt(1) == '=') {
                        cube[] c = {new cube()};
                        if (cubeFromColorsOnSquares(responder, query.substring(2), c)) {
                            responder.message("thread count: " + THREAD_COUNT);
                            if (c[0].equals(csolved))
                                responder.message("already solved");
                            else
                                cubeSearcher.searchMoves(c[0], mode, responder);
                        }
                    } else {
                        responder.message("invalid parameter");
                    }
                }
            } finally {
                solverLock.unlock();
            }
            System.out.println("solver end");
        } else {
            exchange.getResponseHeaders().add("Content-type", "text/plain");
            byte[] resp = "setup: the solver is busy\n".getBytes();
            exchange.sendResponseHeaders(200, resp.length);
            try (OutputStream fdReq = exchange.getResponseBody()) {
                fdReq.write(resp);
            }
            System.out.println("solver busy");
        }
    }

    private static void getFile(HttpExchange exchange) throws IOException {
        String fname = exchange.getRequestURI().getPath().substring(1);
        if (fname.isEmpty()) fname = "cube.html";
        InputStream fp = Thread.currentThread().getContextClassLoader().getResourceAsStream(fname);
        if (fp != null) {
            try {
                int dotPos = fname.lastIndexOf('.');
                String fnameExt = (dotPos < 0) ? "txt" : fname.substring(dotPos+1);
                String contentType = "application/octet-stream";
                if (fnameExt.equalsIgnoreCase("html")) contentType = "text/html; charset=utf-8";
                else if (fnameExt.equalsIgnoreCase("css")) contentType = "text/css";
                else if (fnameExt.equalsIgnoreCase("js"))  contentType = "text/javascript";
                else if (fnameExt.equalsIgnoreCase("txt")) contentType = "text/plain";
                exchange.getResponseHeaders().add("Content-type", contentType);
                exchange.sendResponseHeaders(200, 0);
                try (OutputStream fdReq = exchange.getResponseBody()) {
                    fp.transferTo(fdReq);
                }
            } catch (IOException exc) {
                System.out.println("error: socket write");
            } finally {
                fp.close();
            }
        } else {
            exchange.sendResponseHeaders(404, -1);
        }
    }

    private static class ConnectionHandler implements HttpHandler {
        private final CubeSearcher cubeSearcher;
        ConnectionHandler(CubeSearcher cs) { cubeSearcher = cs; }

        @Override
        public void handle(HttpExchange exchange) {
            try {
                String method = exchange.getRequestMethod();
                var uri = exchange.getRequestURI();
                System.out.println(method + " " + uri);
                if (!method.equalsIgnoreCase("get")) {
                    exchange.sendResponseHeaders(405, -1);
                    return;
                }
                if (uri.getPath().equals("/") && uri.getQuery() != null)
                    processCubeReq(exchange, cubeSearcher);
                else
                    getFile(exchange);
            } catch (Exception exc) {
                System.out.println("exception: " + exc);
                exc.printStackTrace();
            }
        }
    }

    public static void runServer(int depthMax, boolean useReverse) throws IOException {
        CubeSearcher cubeSearcher = new CubeSearcher(depthMax, useReverse);
        HttpServer server = HttpServer.create(new InetSocketAddress(8080), 0);
        server.createContext("/", new ConnectionHandler(cubeSearcher));
        server.start();
    }
}

File: src/main/kotlin/cubesrv/CServer.kt
package cubesrv

import com.sun.net.httpserver.HttpExchange
import com.sun.net.httpserver.HttpServer
import com.sun.net.httpserver.HttpHandler
import java.io.File
import java.io.FileInputStream
import java.io.InputStream
import java.io.IOException
import java.io.OutputStream
import java.net.InetSocketAddress
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.sync.Mutex

private class SocketChunkedResponder(val m_fdReq : OutputStream) : Responder() {
    private val respMutex = Mutex()
    override fun handleMessage(mt : MessageType, msg : String) {
        val prefix : String = when(mt) {
            MessageType.MT_UNQUALIFIED -> ""
            MessageType.MT_PROGRESS    -> "progress: "
            MessageType.MT_MOVECOUNT   -> "moves: "
            MessageType.MT_SOLUTION    -> "solution:"
        }
        val respBuf = "%s%s\n".format(prefix, msg)
        runBlocking {
            respMutex.lock()
            try {
                m_fdReq.write(respBuf.toByteArray())
                m_fdReq.flush()
            }finally{
                respMutex.unlock()
            }
        }
    }
}

private val solverMutex = Mutex()
private fun processCubeReq(exchange : HttpExchange, cubeSearcher : CubeSearcher) {
    val query = exchange.getRequestURI().getQuery()
    if( query.startsWith("stop") ) {
        runBlocking {
            ProgressBase.requestStop()
        }
        exchange.sendResponseHeaders(200, -1)
        return
    }
    if( solverMutex.tryLock() ) {
        runBlocking {
            ProgressBase.requestRestart()
        }
        val headers = exchange.getResponseHeaders()
        headers.add("Content-type", "text/plain")
        exchange.sendResponseHeaders(200, 0)
        val fdReq = exchange.getResponseBody()
        val responder = SocketChunkedResponder(fdReq)
        val mode  : String = query.substring(0, 1)
        if( query[1] == '=' ) {
            val c = arrayOf(cube())
            if( cubeFromColorsOnSquares(responder, query.substring(2), c) ) {
                responder.message("thread count: $THREAD_COUNT")
                if( c[0] == csolved )
                    responder.message("already solved")
                else
                    cubeSearcher.searchMoves(c[0], mode, responder)
            }
        }else{
            responder.message("invalid parameter")
        }
        fdReq.close()
        solverMutex.unlock()
        println("solver end")
    }else{
        val headers = exchange.getResponseHeaders()
        headers.add("Content-type", "text/plain")
        val resp = "setup: the solver is busy\n".toByteArray()
        exchange.sendResponseHeaders(200, resp.size.toLong())
        val fdReq = exchange.getResponseBody()
        fdReq.write(resp)
        fdReq.close()
        println("solver busy")
    }
}

private fun getFile(exchange : HttpExchange) {
    var fname = exchange.getRequestURI().getPath().substring(1)
    if( fname.isEmpty() )
        fname = "cube.html"
    Thread.currentThread().getContextClassLoader().getResourceAsStream(fname).use { fp ->
        if( fp != null ) {
            val dotPos = fname.lastIndexOf(".")
            val fnameExt = if( dotPos < 0 ) "txt" else fname.substring(dotPos+1)
            var contentType = "application/octet-stream"
            if( fnameExt.equals("html", ignoreCase = true) )
                contentType = "text/html; charset=utf-8"
            else if( fnameExt.equals("css", ignoreCase = true) )
                contentType = "text/css"
            else if( fnameExt.equals("js", ignoreCase = true) )
                contentType = "text/javascript"
            else if( fnameExt.equals("txt", ignoreCase = true) )
                contentType = "text/plain"
            val headers = exchange.getResponseHeaders()
            headers.add("Content-type", contentType)
            exchange.sendResponseHeaders(200, 0)
            val fdReq = exchange.getResponseBody()
            try{
                fp.copyTo(fdReq)
                fdReq.close()
            }catch(exc : IOException) {
                println("error: socket write")
                return
            }
        }else{
            exchange.sendResponseHeaders(404, -1)
        }
    }
}

private class ConnectionHandler(val cubeSearcher : CubeSearcher) : HttpHandler {
    override fun handle(exchange : HttpExchange) {
        try {
            val method = exchange.getRequestMethod()
            val uri = exchange.getRequestURI()
            println("$method $uri")
            if( !method.equals("get", ignoreCase = true) ) {
                exchange.sendResponseHeaders(405, -1)
                return
            }
            if( uri.getPath().equals("/") && uri.getQuery() != null )
                processCubeReq(exchange, cubeSearcher)
            else
                getFile(exchange)
        }catch(exc : Exception) {
            println("exception: $exc")
            exc.printStackTrace()
        }
    }
}

fun runServer(depthMax : Int, useReverse : Boolean) {
    val cubeSearcher = CubeSearcher(depthMax, useReverse)
    val server = HttpServer.create(InetSocketAddress(8080), 0);
    server.createContext("/", ConnectionHandler(cubeSearcher));
    server.start();
}
